Add tree-based code generation

VladimirNik opened the pull request:
New printer (ParsedTreePrinter) is intended for unattributed trees and generates the code based on the passed AST.
Result code can be later compiled by scalac retaining the same meaning.
ParsedTreePrinter is based on [Sprinter project](http://vladimirnik.github.io/sprinter/).

VladimirNik committed:
Variance annotations printing

def printTypeParams is modified. Tests are updated.

VladimirNik committed:
val showOuterTests is removed

val showOuterTests wasn't used in Printers

xeno-by commented:
Yay! Happy New Year!

xeno-by commented:
It's great that there's a cleanup commit - it really helps minimizing and understanding the changes. To be really precise, I'd also move the following diffs from the toCode commit to cleanup:
- https://github.com/VladimirNik/scala/commit/ba5b310412e16c32bcea5c37b5bb84dfc11ecbb8#diff-a47b5ef76eb0b2ada3a781e1584539a6L137
- https://github.com/VladimirNik/scala/commit/ba5b310412e16c32bcea5c37b5bb84dfc11ecbb8#diff-a47b5ef76eb0b2ada3a781e1584539a6L181
- https://github.com/VladimirNik/scala/commit/ba5b310412e16c32bcea5c37b5bb84dfc11ecbb8#diff-a47b5ef76eb0b2ada3a781e1584539a6L198
- https://github.com/VladimirNik/scala/commit/ba5b310412e16c32bcea5c37b5bb84dfc11ecbb8#diff-a47b5ef76eb0b2ada3a781e1584539a6L214
- https://github.com/VladimirNik/scala/commit/ba5b310412e16c32bcea5c37b5bb84dfc11ecbb8#diff-a47b5ef76eb0b2ada3a781e1584539a6L266

xeno-by commented:
Great job! It's very cool to see this kind of effort, and I'm looking forward to having `toCode` in trunk! Review @retronym /cc @densh

retronym commented:
Similar functionality exists over in scala-refactoring:

https://github.com/scala-ide/scala-refactoring/blob/master/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/sourcegen/PrettyPrinter.scala

Could you compare please make sure that your test suite covers at least the cases in [PrettyPrinterTest](https://github.com/scala-ide/scala-refactoring/blob/master/org.scala-refactoring.library/src/test/scala/scala/tools/refactoring/tests/sourcegen/PrettyPrinterTest.scala) ?

/cc @misto

VladimirNik commented:
@retronym Yes, as far as I know tree printing in scala-refactoring is targeted on typed trees produced by presentation compiler. Our target in Sprinter was to generate semantically equivalent code using mainly tree structure (we used only dependency on scala.reflect for our main printer), that's why there's a lot of desugaring in result code and we plan to improve it in the future.

I added to my test suite new tests with some code snippets from scala-refactoring tests. Also to test new ParsedTreePrinter I used scalaz (scalaz sources were regenerated using [printPlugin](https://github.com/VladimirNik/printPlugin), recompiled, and then all project's tests were passed successfully).

VladimirNik committed:
toCode is added to Printers

1. Printers API is updated (def toCode)
2. ParsedTreePrinter is added to internal.Printers. ParsedTreePrinter generates the code for passed unattributed trees.
Generated code can be later compiled by scalac retaining the same meaning.

VladimirNik committed:
Tests for ParsedTreePrinter

VladimirNik committed:
Printers code refactoring and cleanup

VladimirNik commented:
Version with fixes from comments is added. Commits are updated.

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -175,25 +187,136 @@ trait Printers extends api.Printers { self: SymbolTable =>
         case anns => anns
       }
-      annots foreach (annot => print("@"+annot+" "))
+      annots foreach (annot => print("@" + annot + " "))
```
`s"@$annot "` ?

retronym replied:
More to the point, does `annotation.toString` always produce re-typecheckable code? What about back-ticks for special characters? How are the arguments to the annotation printed?

VladimirNik replied:
@densh Code fixed
@retronym It's a version of printAnnotations method for TreePrinter. ParsedTreePrinter overrides this [method](https://github.com/VladimirNik/scala/blob/3989227e45a8b7f56da59b7ce674706b1e99279b/src/reflect/scala/reflect/internal/Printers.scala#L663) and process back-ticks for special characters. In the case of TreePrinter arguments to the annotation are printed as a part of tree because toString results in invocation of TreePrinter's printTree(annot).

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -459,11 +525,496 @@ trait Printers extends api.Printers { self: SymbolTable =>
     }
   }
+  
+  //printer for trees after parser and before typer phases
```
Maybe one more space after `//` and before beginning of the comment?

VladimirNik replied:
Fixed

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -459,11 +525,496 @@ trait Printers extends api.Printers { self: SymbolTable =>
+            printBlock(rhs)
+          }
+
+        case imp @ Import(expr, selectors) =>
```
`selectors` not used? use `_` instead?

VladimirNik replied:
Fixed

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -104,9 +102,14 @@ trait Printers extends api.Printers { self: SymbolTable =>
     def printRow(ts: List[Tree], sep: String) { printRow(ts, "", sep, "") }
 
     def printTypeParams(ts: List[TypeDef]) {
```
Procedure syntax is planned to be deprecated in the future, use `def f: Unit = ...` instead.

VladimirNik replied:
Fixed

densh commented:
I see a lot of promise in this work. There are a few low hanging fruits that could be improved by using Syntactic extractors but it would be extremely easy to add them as follow-ups later on.

densh commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -459,11 +525,496 @@ trait Printers extends api.Printers { self: SymbolTable =>
+        case th @ This(qual) =>
+          printThis(th, printedName(qual))
+
+        case Select(qual @ New(tpe), name) =>
```
`tpe` not used, maybe `qual: New` subpattern instead?

VladimirNik replied:
Fixed

retronym commented on src/reflect/scala/reflect/internal/Printers.scala:
```diff
@@ -175,25 +187,136 @@ trait Printers extends api.Printers { self: SymbolTable =>
+    }
+
+    protected def printAnnotated(tree: Annotated)(printArg: => Unit) {
+      val Annotated(Apply(Select(New(tpt), nme.CONSTRUCTOR), args), atree) = tree
```
Does this handle annotations with multiple parameter lists?

VladimirNik replied:
I resolved this issue, new [version](https://github.com/VladimirNik/scala/blob/3989227e45a8b7f56da59b7ce674706b1e99279b/src/reflect/scala/reflect/internal/Printers.scala#L668)  handles Annotated trees that contain annotations with multiple parameter lists.

VladimirNik committed:
Annotated trees processing is modified

1. Problem with multiple parameter lists in annotations is resolved
2. Tests for Annotated trees are added

VladimirNik committed:
Code cleanup based on pull request comments

VladimirNik commented:
@xeno-by @densh @retronym Thanks for the comments! Fixed version is committed.

xeno-by commented:
@VladimirNik You know, I've been thinking now, and how about we turn `Tree.toCode` into `Universe.showCode` in order to be consistent with `show` and `showRaw` (https://github.com/xeno-by/scala/blob/master/src/reflect/scala/reflect/api/Printers.scala#L196)?

VladimirNik committed:
toCode renamed to showCode

VladimirNik commented:
@xeno-by I agree, the only thing - before we had toCode in Printers (and could access it from Universe, not Tree). In my last commit toCode(tree: Tree) is changed to showCode(tree: Tree)

adriaanm commented:
awesome! could one of the reviewers sign off with an LGTM? (remember, it has to be the first word of your comment in order to count)

xeno-by commented:
LGTM

xeno-by commented:
Shall we? :)

densh commented:
:cake: to the @VladimirNik, keep up the good stuff.

adriaanm commented:
Cool stuff indeed, but it breaks on windows unfortunately: https://scala-webapps.epfl.ch/jenkins/job/scala-nightly-windows/1945/console

adriaanm commented:
You can test on windows with https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual-windows/build?delay=0sec

adriaanm commented:
Since we unfortunately don't archive test output on nightlies, I quickly went ahead and capture the relevant log file: https://gist.github.com/adriaanm/8449000 

It looks like the failures are related to the windowsy EOL

VladimirNik commented:
@adriaanm Thank you for log file. Yes, these problems are caused by windows EOL character. Newline processing and test suite are fixed in [#3377](https://github.com/scala/scala/pull/3377).
